### ISSUE TITLE
classif.xgboost - error if early.stop.round is set?

### ISSUE THREAD

**Issue by DanielKuehn87** [mlr-org/mlr#1264]
Not sure what I am doing wrong here:

```
library(mlr)
mlr_task <- pid.task
mlr_learner <- makeLearner("classif.xgboost",
                           par.vals = list(nrounds = 250,
                                           early.stop.round = 25,
                                           maximize = FALSE))

mlr_train <- train(mlr_learner, mlr_task)
```

This results in an error (I tested the mlr CRAN & github version):

> Error in xgb.train(params, dtrain, nrounds, watchlist, verbose = verbose,  : 
> Please set maximize to note whether the model is maximizing the evaluation or not.

**Comment by MariaErdmann** [comment]
Hi DanielKuehn87,
this error message is not related to mlr. It is related to xgboost package itself. A similar issue was reported to the R xgboost team on github: [https://github.com/dmlc/xgboost/issues/916](url), but it seems this is not fixed on CRAN yet.

**Comment by DanielKuehn87** [comment]
Ok, thanks for the link. After using the xgboost version from the drat repo (and updating early.stop.round to early_stopping_rounds in mlr) it works now.

Still a bit weird, since this works with CRAN-xgboost:
```
library(xgboost)
library(mlbench)
data(PimaIndiansDiabetes)

xg_data <- data.matrix(PimaIndiansDiabetes[,1:8])
xg_label <- PimaIndiansDiabetes$diabetes
xg_matrix <- xgb.DMatrix(data = xg_data, 
                         label = xg_label)

xg_watchlist <- list(msr = xg_matrix)

xg_model <- xgb.train(data                = xg_matrix,
                      watchlist           = xg_watchlist,
                      nrounds             = 250, 
                      early.stop.round    = 25,
                      maximize            = FALSE
)

```

**Comment by larskotthoff** [comment]
Hang on, so `early.stop.round` has been renamed in the repo version? We need to update that when it's released on CRAN.

Also, are you saying that you're getting an error with the current development mlr version and the xgboost version on CRAN?

**Comment by DanielKuehn87** [comment]
After reading the xgboost repo docu, it seems like they even changed a bit more: 
1. Missing default is NA now (instead of 0) 
2. Early.stop.round is renamed to early_stopping_rounds
3. Additional parameters save_period, save_name, xgb_model and callbacks were added

Yes, I got the error message above with the latest mlr version on github and xgboost from CRAN. After installing the xgboost repo version it works now.

**Comment by larry77** [comment]
I jump into this thread because I am also experiencing a lot of tiny annoying problems.
      > library(xgboost)
     > library(mlr) 
     > packageVersion("mlr")
     [1] ‘2.9’
     > packageVersion("xgboost")
     [1] ‘0.4.4’

I have the same error about maximize.
Now, if I choose to install from the drat repo of xgboost in order to treat the maximize error, then I have an error with predict as mentioned here

https://github.com/mlr-org/mlr/issues/935

What should I do? It seems that we are dealing with trivial bugs mostly on the xgboost side, but according to the version that I install, I squeeze one bug but I then resuscitate another one.
Any suggestions?

**Comment by larskotthoff** [comment]
mlr version 2.10 (which should be released really soon) should work fine with the CRAN version of xgboost. For now you can use the git version of mlr with the CRAN version of xgboost.

**Comment by larry77** [comment]
Hi!
Well, it does not seem to work on my system.

See my sessionInfo

> sessionInfo()
R version 3.3.1 (2016-06-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Debian GNU/Linux 8 (jessie)

locale:
 [1] LC_CTYPE=en_GB.utf8       LC_NUMERIC=C             
 [3] LC_TIME=en_GB.utf8        LC_COLLATE=en_GB.utf8    
 [5] LC_MONETARY=en_GB.utf8    LC_MESSAGES=en_GB.utf8   
 [7] LC_PAPER=en_GB.utf8       LC_NAME=C                
 [9] LC_ADDRESS=C              LC_TELEPHONE=C           
[11] LC_MEASUREMENT=en_GB.utf8 LC_IDENTIFICATION=C      

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mlr_2.10         ParamHelpers_1.9

loaded via a namespace (and not attached):
 [1] Rcpp_0.12.5      lattice_0.20-33  chron_2.3-47     grid_3.3.1      
 [5] plyr_1.8.4       gtable_0.2.0     backports_1.0.3  magrittr_1.5    
 [9] scales_0.4.0     ggplot2_2.1.0    stringi_1.1.2    data.table_1.9.6
[13] parallelMap_1.3  Matrix_1.2-6     checkmate_1.8.1  xgboost_0.4-4   
[17] splines_3.3.1    tools_3.3.1      stringr_1.0.0    munsell_0.4.3   
[21] survival_2.39-4  parallel_3.3.1   colorspace_1.2-6 BBmisc_1.10     

and the versions of the mlr and xgboost packages

> packageVersion("mlr")
[1] ‘2.10’


> packageVersion("xgboost")
[1] ‘0.4.4’

The example script 

library(mlr)
mlr_task <- pid.task
mlr_learner <- makeLearner("classif.xgboost",
                           par.vals = list(nrounds = 250,
                                           early.stop.round = 25,
                                           maximize = FALSE))

mlr_train <- train(mlr_learner, mlr_task)


returns the following error.

Error in xgb.train(params, dtrain, nrounds, watchlist, verbose = verbose,  : 
  For early stopping you need at least one set in watchlist.
Timing stopped at: 0.028 0 0.013 

Any ideas?

**Comment by larskotthoff** [comment]
This is a different error. We'll have a look at it.

**Comment by ja-thomas** [comment]
This seems to be a bug in xgboost https://github.com/dmlc/xgboost/issues/295

It seems to be fixed but it still does not work on the latest build.

When setting `verbose = 1` it works for the dev version of xgboost but on the CRAN version I'll get a different error I do not really understand. What happens is that when `early.stop.round` is set (and `verbose = 1`, which is required due to the bug). The `maximize` argument isn't accepted anymore, e.g.,

```
> train(makeLearner("classif.xgboost", nrounds = 250, early.stop.round = 25), pid.task)
Error in xgb.train(params, dtrain, nrounds, watchlist, verbose = verbose,  : 
  For early stopping you need at least one set in watchlist.
Timing stopped at: 0.028 0 0.018 
> train(makeLearner("classif.xgboost", nrounds = 250, early.stop.round = 25, verbose = 1), pid.task)
Error in xgb.train(params, dtrain, nrounds, watchlist, verbose = verbose,  : 
  Please set maximize to note whether the model is maximizing the evaluation or not.
Timing stopped at: 0.024 0 0.019 
> train(makeLearner("classif.xgboost", nrounds = 250, early.stop.round = 25, verbose = 1, maximize = TRUE), pid.task)
Error in xgb.train(params, dtrain, nrounds, watchlist, verbose = verbose,  : 
  Please set maximize to note whether the model is maximizing the evaluation or not.
Timing stopped at: 0.024 0 0.016 
```

The second and third call would work with the latest xgboost version....

**Comment by larry77** [comment]
It is a very annoying situation because what you mention is unfortunately not the end of the story.
See if you agree with me (I am using the stable version of mlr)
1) if I use the stable version of xgboost, then I end up in any case in the watchlist nightmare described earlier in this thread
2) if I use the development version of xgboost, things do not really work either, see for instance the code below

    library(mlr)
    task <- pid.task


    lrn <- makeLearner("classif.xgboost",par.vals = list(nrounds = 1000,
    early.stop.round = 10, maximize = FALSE))



    ps = makeParamSet(
    makeIntegerParam("max_depth", lower = 7, upper = 10),
    makeIntegerParam("min_child_weight", lower = 0, upper = 3),
    makeNumericParam("eta", lower = 0.02, upper = 0.08),
    makeNumericParam("colsample_bytree", lower = 3, upper = 5, trafo = function(x) x/5),
    makeNumericParam("subsample", lower = 3, upper = 5, trafo = function(x) x/5)
    )


    rdesc = makeResampleDesc("RepCV", folds =5, reps = 5, predict="test")

     ctrl =  makeTuneControlRandom(maxit = 45)


    res = tuneParams(lrn, task = task, resampling = rdesc, par.set = ps,
     control = ctrl, measures = logloss,  show.info=T   )


Which on my machine returns

    Error: 'predict' is not an exported object from 'namespace:xgboost'
i.e. bug 

https://github.com/mlr-org/mlr/issues/935


I suppose these are problems essentially on the xgboost side, but the sad conclusion is that I do not know how to train an xgboost model with stopping conditions under mlr right now.
Any ideas?

**Comment by ja-thomas** [comment]
Hi Larry,

We have already adapted the xgboost learner to the latest build in https://github.com/mlr-org/mlr/tree/xgboost_latest 

When you use this branch and the latest xgboost build you should be good to go. This is not very thoroughly tested, so if you come across errors please tell me.

**Comment by larry77** [comment]
Mmmmhhhhh.... I installed the development version of mlr and both with the xgboost from cran and with the drat version, I always end up with a watchlist issue

    Error in init(env) : 
    For early stopping, watchlist must have at least one element
    In addition: Warning message:
    'early.stop.round' is deprecated.
    Use 'early_stopping_rounds' instead.
    See help("Deprecated") and help("xgboost-deprecated").

**Comment by larry77** [comment]
Hello,
Any progress on this issue? It basically prevents one from using mlr+xgboost and early stopping (and that *is* really a useful feature).

**Comment by khotilov** [comment]
I've submitted a fix in #1375.

However, note that using xgboost's `early_stopping_rounds` is not really useful within the mlr framework, since it would only use the default evaluation metric results in a *training* set (only when `verbose` is set), and so far I don't see any good way to engage with xgboost's `watchlist` functionality from mlr that would not be an ugly hack.

**Comment by larry77** [comment]
I see and you have a point. It does not make much sense to potentially mix different metrics on training and test sets. Thanks for the clarification! So, final question: let's say I let go of the early stopping thing and I focus on the "predict" error. If I try the latest mlr and latest xgboost, it is still there. When will your fix be part of xgboost? If I am misunderstanding something (e.g. I should not use the development version of both mlr and xgboost) please advise me because I must be drowning in a glass of water.

**Comment by ja-thomas** [comment]
The bug that we have to set `verbose = 1` for `early_stopping_rounds` to work is still present in xgboost version 0.62

**Comment by rwarnung** [comment]
@khotilov I like the early stopping in xgboost and when I use xgboost directly I use another set than the train set to determine the stopping time.

At the moment we can not set the parameter watchlist. In the original xgboost package this is a list that has at least one element resulting from a call to xgb.DMatrix.

Something like

`my.pars$watchlist = list(dtest = xgb.DMatrix(data = test.data, data = test.labels))`

Introducing a parameter watchlist would solve the issue, wouldn't it? Similar to the parameter `eval.data` that is used in the recent implementation of the learner `classif.mxff` that uses the mxnet package.
Am I missing something here? Thank you!

**Comment by khotilov** [comment]
If the mlr authors don't mind `watchlist` evaluation data passing through LearnerParams, then it shouldn't be a problem. I can make a PR this weekend.

**Comment by rwarnung** [comment]
@khotilov this would be great. If it works for `classif.mxff` then it should/could be ok for xgboost too. In my mind early stopping is simply necessary to prevent overfitting in a task with a lot of noise.
